File: embedded/bun/runtime/rcon.ts
// RCON protocol client using Bun native TCP sockets.
// Implements the Source RCON protocol: https://developer.valvesoftware.com/wiki/Source_RCON_Protocol

import type { Socket } from "bun";

const PACKET_TYPE_AUTH = 3;
const PACKET_TYPE_AUTH_RESPONSE = 2;
const PACKET_TYPE_COMMAND = 2;
const PACKET_TYPE_RESPONSE = 0;

interface PendingRequest {
  resolve: (value: string) => void;
  reject: (reason: Error) => void;
  data: Buffer;
}

export class RconClient {
  private socket: Socket | null = null;
  private requestId = 0;
  private pending = new Map<number, PendingRequest>();
  private authenticated = false;
  private buffer = Buffer.alloc(0);

  constructor(
    private host: string,
    private port: number,
    private password: string,
  ) {}

  async connect(): Promise<void> {
    return new Promise<void>((resolve, reject) => {
      const self = this;
      let authResolve = resolve;
      let authReject = reject;

      Bun.connect({
        hostname: self.host,
        port: self.port,
        socket: {
          open(socket) {
            self.socket = socket;
            // Send auth packet
            const id = self.nextId();
            const packet = self.encodePacket(id, PACKET_TYPE_AUTH, self.password);
            self.pending.set(id, {
              resolve: () => {
                self.authenticated = true;
                authResolve();
              },
              reject: authReject,
              data: Buffer.alloc(0),
            });
            socket.write(packet);
          },
          data(socket, data) {
            self.buffer = Buffer.concat([self.buffer, Buffer.from(data)]);
            self.processBuffer();
          },
          error(socket, error) {
            authReject(error);
          },
          close() {
            self.socket = null;
            self.authenticated = false;
            // Reject all pending requests
            for (const [, req] of self.pending) {
              req.reject(new Error("RCON connection closed"));
            }
            self.pending.clear();
          },
        },
      });
    });
  }

  async command(cmd: string): Promise<string> {
    if (!this.socket || !this.authenticated) {
      throw new Error("RCON not connected");
    }

    return new Promise<string>((resolve, reject) => {
      const id = this.nextId();
      const packet = this.encodePacket(id, PACKET_TYPE_COMMAND, cmd);
      this.pending.set(id, { resolve, reject, data: Buffer.alloc(0) });
      this.socket!.write(packet);
    });
  }

  disconnect(): void {
    if (this.socket) {
      this.socket.end();
      this.socket = null;
    }
    this.authenticated = false;
  }

  get isConnected(): boolean {
    return this.authenticated && this.socket !== null;
  }

  private nextId(): number {
    return ++this.requestId;
  }

  private encodePacket(id: number, type: number, body: string): Buffer {
    const bodyBuf = Buffer.from(body, "utf-8");
    // Packet: 4 (size) + 4 (id) + 4 (type) + body + 2 (null terminators)
    const size = 4 + 4 + bodyBuf.length + 2;
    const packet = Buffer.alloc(4 + size);
    packet.writeInt32LE(size, 0);
    packet.writeInt32LE(id, 4);
    packet.writeInt32LE(type, 8);
    bodyBuf.copy(packet, 12);
    // Two null bytes at the end (body terminator + packet terminator)
    packet[12 + bodyBuf.length] = 0;
    packet[13 + bodyBuf.length] = 0;
    return packet;
  }

  private processBuffer(): void {
    while (this.buffer.length >= 4) {
      const size = this.buffer.readInt32LE(0);
      const totalPacketSize = 4 + size;

      if (this.buffer.length < totalPacketSize) break;

      const id = this.buffer.readInt32LE(4);
      const type = this.buffer.readInt32LE(8);
      const body = this.buffer.subarray(12, 12 + size - 10).toString("utf-8");

      this.buffer = this.buffer.subarray(totalPacketSize);

      // Auth response
      if (type === PACKET_TYPE_AUTH_RESPONSE) {
        const req = this.pending.get(id) ?? this.pending.get(id - 1);
        if (req) {
          this.pending.delete(id);
          this.pending.delete(id - 1);
          if (id === -1) {
            req.reject(new Error("RCON authentication failed"));
          } else {
            req.resolve(body);
          }
        }
        continue;
      }

      // Command response
      if (type === PACKET_TYPE_RESPONSE) {
        const req = this.pending.get(id);
        if (req) {
          this.pending.delete(id);
          req.resolve(body);
        }
      }
    }
  }
}

File: internal/container/rcon.go
package container

import (
	"context"
	"encoding/binary"
	"fmt"
	"io"
	"net"
	"sync"
	"sync/atomic"
	"time"
)

// Source RCON protocol packet types.
const (
	packetTypeAuth int32 = 3

	// packetTypeAuthResponse and packetTypeCommand both use value 2 per the
	// Source RCON protocol spec. Callers distinguish them by context (auth
	// phase vs command phase) and request ID.
	packetTypeAuthResponse int32 = 2
	packetTypeCommand      int32 = 2

	packetTypeResponse int32 = 0

	// maxRCONBodySize is the maximum size of an RCON response body in bytes.
	maxRCONBodySize = 4096
)

// RCONClient implements the Source RCON protocol for communicating with a
// Minecraft server. It supports concurrent Command calls after Connect
// returns, but Connect and Close must be externally synchronized.
type RCONClient struct {
	addr     string
	password string
	conn     net.Conn
	mu       sync.Mutex
	reqID    atomic.Int32
}

// NewRCONClient creates a new RCON client for the given address and password.
func NewRCONClient(addr, password string) *RCONClient {
	return &RCONClient{
		addr:     addr,
		password: password,
	}
}

// Connect dials the RCON server and authenticates.
func (r *RCONClient) Connect(ctx context.Context) error {
	r.mu.Lock()
	defer r.mu.Unlock()

	dialer := net.Dialer{Timeout: 5 * time.Second}
	conn, err := dialer.DialContext(ctx, "tcp", r.addr)
	if err != nil {
		return fmt.Errorf("rcon dial: %w", err)
	}
	r.conn = conn

	// Apply a deadline for the auth handshake so we don't block forever
	// if the server accepts the connection but never responds.
	authDeadline := time.Now().Add(5 * time.Second)
	if d, ok := ctx.Deadline(); ok && d.Before(authDeadline) {
		authDeadline = d
	}
	_ = conn.SetDeadline(authDeadline)
	defer func() { _ = conn.SetDeadline(time.Time{}) }()

	// Authenticate.
	id := r.nextID()
	if err := r.writePacket(id, packetTypeAuth, r.password); err != nil {
		_ = conn.Close()
		r.conn = nil
		return fmt.Errorf("rcon auth write: %w", err)
	}

	respID, respType, _, err := r.readPacket()
	if err != nil {
		_ = conn.Close()
		r.conn = nil
		return fmt.Errorf("rcon auth read: %w", err)
	}

	// Minecraft sends an auth response with the request ID on success,
	// or -1 on failure.
	if respType == packetTypeAuthResponse && respID == -1 {
		_ = conn.Close()
		r.conn = nil
		return fmt.Errorf("rcon authentication failed")
	}

	// Some servers send an empty command-response packet before the real
	// auth response. Only read the second packet when the first was NOT
	// an auth response (and therefore not a failure).
	if respType != packetTypeAuthResponse {
		respID, respType, _, err = r.readPacket()
		if err != nil {
			_ = conn.Close()
			r.conn = nil
			return fmt.Errorf("rcon auth read (2nd): %w", err)
		}
		if respType == packetTypeAuthResponse && respID == -1 {
			_ = conn.Close()
			r.conn = nil
			return fmt.Errorf("rcon authentication failed")
		}
	}

	return nil
}

// Command sends a command and returns the response body.
func (r *RCONClient) Command(ctx context.Context, cmd string) (string, error) {
	r.mu.Lock()
	defer r.mu.Unlock()

	if r.conn == nil {
		return "", fmt.Errorf("rcon not connected")
	}

	if deadline, ok := ctx.Deadline(); ok {
		_ = r.conn.SetDeadline(deadline)
		defer func() { _ = r.conn.SetDeadline(time.Time{}) }()
	}

	id := r.nextID()
	if err := r.writePacket(id, packetTypeCommand, cmd); err != nil {
		return "", fmt.Errorf("rcon command write: %w", err)
	}

	respID, _, body, err := r.readPacket()
	if err != nil {
		return "", fmt.Errorf("rcon command read: %w", err)
	}
	if respID != id {
		return "", fmt.Errorf("rcon response ID mismatch: got %d, want %d", respID, id)
	}

	return body, nil
}

// Close closes the underlying connection.
func (r *RCONClient) Close() error {
	r.mu.Lock()
	defer r.mu.Unlock()

	if r.conn == nil {
		return nil
	}
	err := r.conn.Close()
	r.conn = nil
	return err
}

func (r *RCONClient) nextID() int32 {
	return r.reqID.Add(1)
}

func (r *RCONClient) writePacket(id, pktType int32, body string) error {
	bodyBytes := []byte(body)
	// Packet layout: 4 (size) + 4 (id) + 4 (type) + body + 2 (null terminators)
	size := int32(4 + 4 + len(bodyBytes) + 2)
	buf := make([]byte, 4+size)
	binary.LittleEndian.PutUint32(buf[0:4], uint32(size))
	binary.LittleEndian.PutUint32(buf[4:8], uint32(id))
	binary.LittleEndian.PutUint32(buf[8:12], uint32(pktType))
	copy(buf[12:], bodyBytes)
	buf[12+len(bodyBytes)] = 0
	buf[13+len(bodyBytes)] = 0

	_, err := r.conn.Write(buf)
	return err
}

func (r *RCONClient) readPacket() (id, pktType int32, body string, err error) {
	// Read the 4-byte size prefix.
	var sizeBuf [4]byte
	if _, err := io.ReadFull(r.conn, sizeBuf[:]); err != nil {
		return 0, 0, "", err
	}
	size := int32(binary.LittleEndian.Uint32(sizeBuf[:]))
	if size < 10 || size > maxRCONBodySize+10 {
		return 0, 0, "", fmt.Errorf("rcon packet size out of range: %d", size)
	}

	payload := make([]byte, size)
	if _, err := io.ReadFull(r.conn, payload); err != nil {
		return 0, 0, "", err
	}

	id = int32(binary.LittleEndian.Uint32(payload[0:4]))
	pktType = int32(binary.LittleEndian.Uint32(payload[4:8]))
	// Body is everything between the type field and the two null terminators.
	bodyLen := size - 10
	if bodyLen > 0 {
		body = string(payload[8 : 8+bodyLen])
	}

	return id, pktType, body, nil
}
